### DISCUSSION TITLE
How to Compute Area Fractions of Variants (child phase) for a Single Grain (parent phase)

### DISCUSSION THREAD

**Discussion by Rhalder322** [ORTools4MTEX/ORTools#69] [Q&A] [answered]
Hello Frank,
I would like to compute area fractions of 12 variants for a single grain for Ti alloys. This line of script "job.calcVariants" provides area fractions of variants for the entire map. When I tried this line of scripts, I encountered errors as follows:

job.grains("781").calcVariants     # 781 was one of the grainIds
_Unrecognized method, property, or field
'calcVariants' for class 'grain2d'.

Error in indexing (line 83)
  [varargout{1:nargout}] = builtin('subsref',grains,s);_

So, could you please help me figure out how I can estimate area fractions of variants for a specific single parent grain?

**Comment by AzdiarGazder** [comment]
Hi @Rhalder322,

I've used the grainId from [ORTools_example02](https://github.com/ORTools4MTEX/ORTools/blob/develop/ORTools_example02.m) and snippets from the [plotStack](https://github.com/ORTools4MTEX/ORTools/blob/develop/src/plotStack.m) function to help answer your query:

```matlab
%% Define the parent grain
pGrainId = 110;
pGrain = job.parentGrains(job.parentGrains.id == pGrainId);
pEBSD = job.ebsd(pGrain);
pEBSD = pEBSD(job.csParent);


%% Define the child grain(s)
clusterGrains = job.grainsPrior(job.mergeId == pGrainId);
cGrains = clusterGrains(job.csChild);
cEBSD = job.ebsdPrior(job.ebsdPrior.id2ind(pEBSD.id));
cEBSD = cEBSD(job.csChild);

% Define a list of unique variants
uniqueVariantIds = 1:max(cGrains.variantId);

% Create a logical matrix indicating where each variant id appears in the
% variable 'cGrains.variantId'
logicalMatrix = bsxfun(@eq, cGrains.variantId(:), uniqueVariantIds(:)');

% Calculate grain area fractions of each variant using the logical matrix
areaFractions = sum(logicalMatrix .* cGrains.area(:), 1) / sum(cGrains.area);
```

<ins>**Lastly, and importantly, if an explanation from the maintainers resolves your issue, please press the "Mark as answer" button on the appropriate response. Your assistance in this regard enables us to close discussions as complete. **</ins>

Hope this helps.

Warm regards,
Azdi

**Reply by Rhalder322** [reply]
Q1. I have a follow up question. 

plot(pGrain) => did not maintain the same IPF color key.  

when I tried the following scripts, I got an error message.
parentIPFkey = plotMap_IPF_p2c(pGrain,vector3d.Z,'linewidth',1.5,'parent');
figure; 
plot(parentIPFkey);

**How can I plot pGrain such that it's color schema matches with the IPF key of the reconstructed map?**

Q2. How to generate the variants map for a single grain using the following function?
plotMap_variants()

Q3. Additionally, if I want to highlight only certain chosen variants in the map, how do I do it?

**Reply by AzdiarGazder** [reply, marked as answer]
Hi @Rhalder322,

I believe you're asking 3 follow-up questions, not 1. :)

From your questions, it appears you have not watched the tutorial videos or gone through any of the example scripts to help you find the answers you are looking for.
So, as a first step, I'm advising you to please watch all the tutorial videos.
While doing that, please click on the functions within the example scripts to see how they are coded.
This will help you understand ORTools and more importantly, it will teach you how to use snippets of code to calculate or plot anything specific you have in mind.

Of course, if a function does not work as intended or if you need some additional functionality that is not part of ORTools, then we're always happy to help out.

Here are the answers to your questions in order of appearance.

1. To plot parent IPF maps:

**Method 1:**
As shown in the example files, you can use the following command after reconstruction is completed.
```matlab
grainClick(job,'grains');
```
This function automatically enables a number of grain-based plots without any further scripting needed from you.

**Method 2:**
Alternatively, and as I mentioned in my earlier reply, please have a look at the [plotStack](https://github.com/ORTools4MTEX/ORTools/blob/develop/src/plotStack.m) function to help you manually plot details for single grains.
```matlab
% Define the parent grain IPF notation
ipfKeyParent = ipfHSVKey(job.csParent);
ipfKeyParent.inversePoleFigureDirection = xvector;

% Plot the parent IPF map
cbsParent = ipfKeyParent.orientation2color(pGrain.meanOrientation);
plot(pGrain,cbsParent);
```

2. To generate the variants map for a single grain:
Please refer to Method 1 (which plots this information automatically) or Method 2 to do it manually.

3. To highlight only certain variant ids in the map:
Similar to my earlier reply, I've used the grainId from [ORTools_example02](https://github.com/ORTools4MTEX/ORTools/blob/develop/ORTools_example02.m) and snippets from the [plotStack](https://github.com/ORTools4MTEX/ORTools/blob/develop/src/plotStack.m) function to help answer your query:
```matlab
%% Define the parent grain
pGrainId = 110;
pGrain = job.parentGrains(job.parentGrains.id == pGrainId);
pEBSD = job.ebsd(pGrain);
pEBSD = pEBSD(job.csParent);


%% Define the child grain(s)
clusterGrains = job.grainsPrior(job.mergeId == pGrainId);
cGrains = clusterGrains(job.csChild);
cEBSD = job.ebsdPrior(job.ebsdPrior.id2ind(pEBSD.id));
cEBSD = cEBSD(job.csChild);

% Define a list of unique variants to plot
uniqueVariantIds = [4 7 10];

% Create a logical matrix indicating where each variant id appears in the
% variable 'cGrains.variantId'
logicalMatrix = any(bsxfun(@eq, cGrains.variantId(:), uniqueVariantIds(:)'),2);

% Plot the unique variant ids
plot(cGrains(logicalMatrix),cGrains.variantId(logicalMatrix))
colormap(jet);
caxis([1 max(cGrains.variantId)]);
```


<ins>**Lastly, and importantly, if an explanation from the maintainers resolves your issue, please press the "Mark as answer" button on the appropriate response. Your assistance in this regard enables us to close discussions as complete. **</ins>

Hope this helps.

Warm regards,
Azdi
